src/drivers/cholqrcp.cc

```diff
@@ -0,0 +1,249 @@
+    long copy_t_dur = 0;
+
+    high_resolution_clock::time_point resize_t_start;
+    high_resolution_clock::time_point resize_t_stop;
+    long resize_t_dur = 0;
+
+    high_resolution_clock::time_point total_t_start;
+    high_resolution_clock::time_point total_t_stop;
+    long total_t_dur = 0;
+    //-------TIMING VARS--------/
+
+    //-------TIMING--------/
+    if(this -> timing) {
+        total_t_start = high_resolution_clock::now();
+    
+        resize_t_start = high_resolution_clock::now();
+    }
+    //-------TIMING--------/
+
+    T* A_dat       = A.data();
+    T* A_hat_dat   = upsize(d * n, this->A_hat);
+    T* tau_dat     = upsize(n, this->tau);
+    J.resize(n);
+    int64_t* J_dat = J.data();
+
+    //-------TIMING--------/
+    if(this -> timing) {
+        resize_t_stop = high_resolution_clock::now();
+        resize_t_dur = duration_cast<microseconds>(resize_t_stop - resize_t_start).count();
+    }
+    
+    //-------TIMING--------/
+    if(this -> timing) {
+        saso_t_start = high_resolution_clock::now();
+    }
+    //-------TIMING--------/
+
+    /*
+    Old way of generating and applying SASO
+
+    struct RandBLAS::sasos::SASO sas;
+    sas.n_rows = d; // > n
+    sas.n_cols = m;
+    sas.vec_nnz = this->nnz;
+    sas.rows = new int64_t[sas.vec_nnz * m];
+    sas.cols = new int64_t[sas.vec_nnz * m];
+    sas.vals = new double[sas.vec_nnz * m];
+    RandBLAS::sasos::fill_colwise(sas, this->seed, 0);
+
+    RandBLAS::sasos::sketch_csccol(sas, n, (double*) A_dat, (double*) A_hat_dat, this->num_threads);
+    */
+
+    RandBLAS::sparse::SparseDist DS = {RandBLAS::sparse::SparseDistName::SASO, d, m, this->nnz};        
+    RandBLAS::sparse::SparseSkOp<T> S(DS, this->seed, 0, NULL, NULL, NULL);        
+    RandBLAS::sparse::fill_sparse(S);
+
+    RandBLAS::sparse::lskges<T, RandBLAS::sparse::SparseSkOp<T>>(blas::Layout::ColMajor, blas::Op::NoTrans, blas::Op::NoTrans, d, n, m, 1.0, S, 0, 0, A.data(), m, 0.0, A_hat_dat, d);
+
+    //-------TIMING--------/
+    if(this -> timing) {
+        saso_t_stop = high_resolution_clock::now();
+        saso_t_dur = duration_cast<microseconds>(saso_t_stop - saso_t_start).count();
+
+        qrcp_t_start = high_resolution_clock::now();
+    }
+    //-------TIMING--------/
+    
+    // QRCP - add failure condition
+    geqp3(d, n, A_hat_dat, d, J_dat, tau_dat);
+
+    //-------TIMING--------/
+    if(this -> timing) {
+        qrcp_t_stop = high_resolution_clock::now();
+        qrcp_t_dur = duration_cast<microseconds>(qrcp_t_stop - qrcp_t_start).count();
+
+        rank_reveal_t_start = high_resolution_clock::now();
+    }
+    //-------TIMING--------/
+
+    // Find rank
+    int64_t k = n;
+    this->rank = k;
+    int i;
+    for(i = 0; i < n; ++i) {
+        if(std::abs(A_hat_dat[i * d + i]) < this->eps) {
+            // "Rank" is k, but the index should be k - 1
+            k = i;
+            this->rank = k;
+            break;
+        }
+    }
```

Comment:
This seems more complicated than necessary. Please switch to the following code (or point out that the following code has an issue):
```
    int rank = n;
    for(int i = 0; i < n; ++i) {
        if(std::abs(A_hat_dat[ i * d + i ]) < this->eps) {
            rank = i;
            break;
        }
    }
    this->rank = rank;
    int k = rank - 1;  // index of the last row/column of R that we'll look at.
```

Reply:
Yes, you are correct. This was convoluted because i have not chenged it since when i was trying to figure out my one-off bug. Also, at the end of the day, k = rank.